llvm/lib/Target/X86/X86ISelLowering.cpp

```diff
@@ -29006,6 +29007,12 @@ static SDValue LowerShiftByScalarImmediate(SDValue Op, SelectionDAG &DAG,
       Op.getOpcode() == ISD::SRA)
     return ArithmeticShiftRight64(ShiftAmt);
 
+  // If we're right logical shifting the MSB to LSB of an all-signbits value
+  // then we can just perform as a mask.
+  if (Op.getOpcode() == ISD::SRL && ShiftAmt == (EltSizeInBits - 1))
+    if (DAG.ComputeNumSignBits(R) == EltSizeInBits)
```

Comment:
We could handle any shift amount, because we are essentially just extracting a mask. How about drop the `ShiftAmt` check and return:
`DAG.getNode(ISD::AND, ..., DAG.getConstant(APInt::getLowBitsSet(bitwidth, bitwidth-ShiftAmount)))`?

Reply:
Also, is there a reason this isn't in `DAGComber`. seems pretty general.

Reply:
In most cases we'd want to avoid the mask load if we can perform an immediate shift.

Reply:
It seems to me most of the benefit comes from the improved analysis.
Maybe doing it generally here and undoing in `X86ISelDAGToDAG` if the constant is single use (although I know we don't do a great job tracking that at the moment) would be ideal?